solr/core/src/java/org/apache/solr/update/processor/TimeRoutedAliasUpdateProcessor.java

```diff
@@ -170,55 +179,35 @@ public void processAdd(AddUpdateCommand cmd) throws IOException {
-      if (routeTimestamp.isAfter(maxFutureTime)) {
-        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
-            "The document's time routed key of " + routeValue + " is too far in the future given " +
-                TimeRoutedAlias.ROUTER_MAX_FUTURE + "=" + timeRoutedAlias.getMaxFutureMs());
-      }
-
-      // Create a new collection?
-      final Instant nextCollTimestamp = timeRoutedAlias.computeNextCollTimestamp(mostRecentCollTimestamp);
-      if (routeTimestamp.isBefore(nextCollTimestamp)) {
-        break; // thus we don't need another collection
-      }
-
-      createCollectionAfter(mostRecentCollName); // *should* throw if fails for some reason but...
-      final boolean updated = updateParsedCollectionAliases();
-      if (!updated) { // thus we didn't make progress...
-        // this is not expected, even in known failure cases, but we check just in case
-        throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
-            "We need to create a new time routed collection but for unknown reasons were unable to do so.");
+    String targetCollection = findCandidateCollectionGivenTimestamp(routeTimestamp, cmd.getPrintableId());
+    try {
+      String preemptiveCreateWindow = timeRoutedAlias.getPreemptiveCreateWindow();
+      if (StringUtils.isBlank(preemptiveCreateWindow) ||                               // default: no pre-create
+          requiresCreateCollection(routeTimestamp, null)) {        // or no collection exists for doc
+        targetCollection = new Maintainer(routeTimestamp, cmd.getPrintableId()).maintain(targetCollection);                      // then do it synchronously.
+      } else if (requiresCreateCollection(routeTimestamp, preemptiveCreateWindow )) {  // else async...
+        String finalTargetCollection = targetCollection;
+        try {
+          asyncMaintainers.computeIfAbsent(timeRoutedAlias.getAliasName(),(alias) ->
+              singleThreadSingleTaskExecutor()).submit(() ->
+              new Maintainer(routeTimestamp, cmd.getPrintableId()).maintain(finalTargetCollection));
+        } catch (RejectedExecutionException e) {
+          // Note: There are some esoteric cases where the pre-create interval is a lot longer than
+          // a single time slice in the alias, and a doc that should cause several creations is blocked by
+          // one that only creates a single collection, but that requires some sort of long pause followed by a 2 doc
+          // burst and even then it only causes a synch create if there's a subsequent pause of more than another
+          // full time slice... These sorts of cases are probably more work than their value. If someone actually
+          // hits those cases frequently enough to notice, we can think about adding some further logic. For now,
+          // this is complicated enough as it is.
+          log.trace("Collection creation rejected, probably some other request has" +
```

Comment:
I'm not yet sure what to think of the scenario above, but note the code that existed before had an outer loop and that helped with edge cases by retrying.